### ISSUE TITLE
fix: Boolean binding to some FAST control attributes does not behave as expected

### ISSUE THREAD

**Issue by rajsite** [microsoft/fast#5320]
# 🐛 Bug Report

Some boolean attributes on FAST controls do not behave how [mdn](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes#boolean_attributes) or [FAST template binding](https://www.fast.design/docs/fast-element/declaring-templates/#attributes) expect boolean attributes to behave.

This issue impacts all boolean attributes configured for a default of true such as dialog's `modal` property discussed in the issue and could potentially be [avoided for picker](https://github.com/microsoft/fast/blob/45f000f038302d31636bccd85cf20a366aad7a0b/packages/web-components/fast-foundation/src/picker/picker.ts#L93)'s `filter-selected` and `filter-query` attributes which is in alpha.

## 💻 Repro or Code Sample

The Fast Dialog's `modal` attribute cannot be set to false as expected with `<fast-dialog></fast-dialog>`. You have to use the unexpected `<fast-dialog modal="false"></fast-dialog>` which does not follow boolean attribute conventions in browser or work with FAST's own boolean attribute template binding.

See the following StackBlitz where trying to bind to the [modal boolean attribute](https://github.com/microsoft/fast/blob/45f000f038302d31636bccd85cf20a366aad7a0b/packages/web-components/fast-foundation/src/dialog/dialog.ts#L22) of Dialog the binding does not work as expected: https://stackblitz.com/edit/fast-boolean-binding-dialog-modal?file=index.ts

## 🤔 Expected Behavior

I expect to be able to use boolean attributes as documented and expected in web applications, where absence of an attribute corresponds to a false value and presence of an attribute corresponds to a true value.

## 😯 Current Behavior

I cannot set some boolean attributes to false with expected attribute conventions. The unexpected attribute strings "false" and "true" do appear to give boolean behavior in these cases.

## 💁 Possible Solution

- Clearly document boolean properties that do not behave normally and must be configured with the strings "true" or "false" instead of normal boolean attribute behavior.

- It would be a breaking API change to correct this behavior and make it behave as expected. One option would be to change these properties from boolean attributes to string attributes where the attribute values "enabled", "true", "", and null correspond to boolean value `true` and the attribute values  "disabled" and "false" correspond to boolean value `false`. Then deprecate usages of the string values other than "enabled" or "disabled" to encourage the use of those attribute strings in docs.

- Require that all future boolean attributes default to the value false. Attributes may need a negated name for good defaults. ie if default modal true is desired then the attribute should instead be called `nonmodal` and correspond to JS property `modal`.

## 🔦 Context

This behavior was not intuitive and something we had to document clearly and make our users aware of using elements built on FastFoundation.

## 🌍 Your Environment

    "@microsoft/fast-element": "^1.4.1",
    "@microsoft/fast-foundation": "^2.12.0",

**Comment by chrisdholt** [comment]
Will be picked up with FAST Foundation vNext work

**Comment by chrisdholt** [comment]
> @chrisdholt @EisenbergEffect In the fast-element-2 branch it looks like there are a couple more instances of a boolean attribute defaulting to true:
> 
> Another in the dialog (trapFocus):
> 
> * https://github.com/microsoft/fast/blob/features/fast-element-2/packages/web-components/fast-foundation/src/dialog/dialog.ts#L51
> 
> And a couple others components:
> 
> * https://github.com/microsoft/fast/blob/features/fast-element-2/packages/web-components/fast-foundation/src/picker/picker.ts#L117
> * https://github.com/microsoft/fast/blob/features/fast-element-2/packages/web-components/fast-foundation/src/picker/picker.ts#L127
> * https://github.com/microsoft/fast/blob/features/fast-element-2/packages/web-components/fast-foundation/src/tabs/tabs.ts#L150
> 
> 1. Will those be updated to default to false as well?
> 2. Is there a type change / or runtime check for the decorator to assert that the value does not default to true to catch that mistake earlier? Or should it be documented like mentioned here: [fix: Boolean binding to some FAST control attributes does not behave as expected #5320 (comment)](https://github.com/microsoft/fast/issues/5320#issuecomment-951333189)

I'll take a look - trap focus seems like it's going to be a difficult one to figure out into false in scenarios where it should "just work". I'll know more about the others when I look.

**Comment by rajsite** [comment]
> I'll take a look - trap focus seems like it's going to be a difficult one to figure out into false in scenarios where it should "just work". I'll know more about the others when I look.

Yea in those scenarios we have inverted the name, ie I could see changing "trap-focus" => "avoid-trapping" so by default it traps.

**Comment by EisenbergEffect** [comment]
We should definitely fix this so that all boolean attributes default to `false`.

**Comment by chrisdholt** [comment]
> We should definitely fix this so that all boolean attributes default to `false`.

Boolean fix for dialog trapFocus is linked above.

**Comment by chrisdholt** [comment]
> We should definitely fix this so that all boolean attributes default to `false`.

Tabs is up at the link above

**Comment by rajsite** [comment]
@chrisdholt @EisenbergEffect  These updates are great! In the vein of potentially breaking changes making it in for FE2 do you think getting noImplicitAny and strictPropertyInitialization can also be targeted for that branch? https://github.com/microsoft/fast/issues/5338

It's been a recurring issue for us that components property types don't align with their values, usually because they are not initialized and can be undefined, ie https://github.com/microsoft/fast/blob/master/packages/web-components/fast-foundation/src/tab/tab.ts#L19

**Comment by chrisdholt** [comment]
> @chrisdholt @EisenbergEffect These updates are great! In the vein of potentially breaking changes making it in for FE2 do you think getting noImplicitAny and strictPropertyInitialization can also be targeted for that branch? #5338
> 
> It's been a recurring issue for us that components property types don't align with their values, usually because they are not initialized and can be undefined, ie https://github.com/microsoft/fast/blob/master/packages/web-components/fast-foundation/src/tab/tab.ts#L19

We'll be making some "progressive" breaks as part of the beta. I'll look for low hanging fruit to see if we could sneak it, but it's coming in hot for Thursday release day.

**Comment by chrisdholt** [comment]
@rajsite I think perhaps one solution we could do for an initial pass here would be - @EisenbergEffect?

```ts
public foo: boolean;
```

```ts
public foo?: boolean;
```

**Comment by rajsite** [comment]
> @rajsite I think perhaps one solution we could do for an initial pass here would be - @EisenbergEffect?
> 
> ```ts
> public foo: boolean;
> ```
> 
> ```ts
> public foo?: boolean;
> ```

For attribute-bound booleans I think it's pretty safe to initialize them to `false` (since that does not change the attribute value in the DOM).

For other types (string, number) making them optional would be a good first step, as the type definition then better aligns with the actual value when accessed at runtime.

**Comment by chrisdholt** [comment]
> > @rajsite I think perhaps one solution we could do for an initial pass here would be - @EisenbergEffect?
> > ```ts
> > public foo: boolean;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```ts
> > public foo?: boolean;
> > ```
> 
> For attribute-bound booleans I think it's pretty safe to initialize them to `false` (since that does not change the attribute value in the DOM).
> 
> For other types (string, number) making them optional would be a good first step, as the type definition then better aligns with the actual value when accessed at runtime.

Agree here - my biggest concern is string initialization, etc and typically because those often *aren't* actually required, where a boolean should almost always have an initial value.

**Comment by rajsite** [comment]
> Agree here - my biggest concern is string initialization, etc and typically because those often _aren't_ actually required, where a boolean should almost always have an initial value.

Makes sense. My ultimate end goal is getting as close as possible to API compatibility with native elements for our fast-foundation based elements. Or not achieving that, having strict types to help our devs out.

So for some properties like the `type` on a vanilla button with html `<button></button>`  returns `'submit'` while the `type` on a fast foundation based button returns `undefined` for the html template `<fast-button></fast-button>` which doesn't align with expectations or [specs](https://html.spec.whatwg.org/multipage/form-elements.html#dom-button-type).

But definitely agree that it takes case-by-case consideration and a good step for string types is marking them optional so in our code bases which are strict we are at least checked by the TS compiler. Then follow-ups can be done to initialize where it makes sense like button type, etc.

**Comment by chrisdholt** [comment]
> > Agree here - my biggest concern is string initialization, etc and typically because those often _aren't_ actually required, where a boolean should almost always have an initial value.
> 
> Makes sense. My ultimate end goal is getting as close as possible to API compatibility with native elements for our fast-foundation based elements. Or not achieving that, having strict types to help our devs out.
> 
> So for some properties like the `type` on a vanilla button with html `<button></button>` returns `'submit'` while the `type` on a fast foundation based button returns `undefined` for the html template `<fast-button></fast-button>` which doesn't align with expectations or [specs](https://html.spec.whatwg.org/multipage/form-elements.html#dom-button-type) (assuming we want to get to api compatibility).
> 
> But definitely agree that it takes case-by-case consideration and a good step for string types is marking them optional so in our code bases which are configured for strict type checking we are at least checked by the TS compiler. Then follow-ups can be done to initialize where it makes sense like button type, etc.

I'm going to go and try and set these for booleans, but I don't think we can turn it off completely. I just ran locally and immediately hit an issue with `id` in Accordion Item - I think because the HTMLElement interface is set to `string` and not `string | undefined/null`. So, the best we can probably do for now is set the boolean values.

**Comment by chrisdholt** [comment]
@EisenbergEffect @rajsite PR is here for setting default values for boolean attrs: https://github.com/microsoft/fast/pull/6023

**Comment by rajsite** [comment]
@chrisdholt do you know the state of #6023? It seems like it was closed a couple weeks ago but I'm having trouble finding a discussion on the outcome.

**Comment by EisenbergEffect** [comment]
I believe that the outcome wast to always make default boolean values `false` so that this issue doesn't happen. However, that's a breaking change so it would need to come in our upcoming major version. @Chrisdholt can correct me if I'm wrong.

**Comment by chrisdholt** [comment]
> @chrisdholt do you know the state of #6023? It seems like it was closed a couple weeks ago but I'm having trouble finding a discussion on the outcome.

Yeah I closed this as it was so far out of date. It should be light work to be honest as these are all "false", I just need to get a branch setup. In some cases, code will likely need to be changed to ensure that some of these which bind to ARIA attributes don't _add_ ARIA where it's not expected (there are a few of these cases). 

We did the work I believe in Foundation already to remove all defaulting to true, so the values all should default to *undefined* right now. We can certainly default to false, but requires what's noted above to be "done".
